// lodash/lodash.js
// Fake lodash implementation for testing
// Simulating a larger utility library

const lodash = {
  // Array methods
  chunk: function(array, size) {
    const result = [];
    for (let i = 0; i < array.length; i += size) {
      result.push(array.slice(i, i + size));
    }
    return result;
  },

  compact: function(array) {
    return array.filter(Boolean);
  },

  concat: function(...arrays) {
    return [].concat(...arrays);
  },

  difference: function(array, ...values) {
    const valuesSet = new Set([].concat(...values));
    return array.filter(item => !valuesSet.has(item));
  },

  drop: function(array, n = 1) {
    return array.slice(n);
  },

  flatten: function(array) {
    return [].concat(...array);
  },

  uniq: function(array) {
    return [...new Set(array)];
  },

  // Collection methods
  map: function(collection, iteratee) {
    return Array.isArray(collection)
      ? collection.map(iteratee)
      : Object.values(collection).map(iteratee);
  },

  filter: function(collection, predicate) {
    return Array.isArray(collection)
      ? collection.filter(predicate)
      : Object.values(collection).filter(predicate);
  },

  reduce: function(collection, iteratee, accumulator) {
    return Array.isArray(collection)
      ? collection.reduce(iteratee, accumulator)
      : Object.values(collection).reduce(iteratee, accumulator);
  },

  forEach: function(collection, iteratee) {
    if (Array.isArray(collection)) {
      collection.forEach(iteratee);
    } else {
      Object.values(collection).forEach(iteratee);
    }
  },

  find: function(collection, predicate) {
    return Array.isArray(collection)
      ? collection.find(predicate)
      : Object.values(collection).find(predicate);
  },

  // Object methods
  assign: function(target, ...sources) {
    return Object.assign(target, ...sources);
  },

  clone: function(value) {
    return JSON.parse(JSON.stringify(value));
  },

  cloneDeep: function(value) {
    return JSON.parse(JSON.stringify(value));
  },

  get: function(object, path, defaultValue) {
    const keys = Array.isArray(path) ? path : path.split('.');
    let result = object;
    for (const key of keys) {
      result = result?.[key];
      if (result === undefined) return defaultValue;
    }
    return result;
  },

  set: function(object, path, value) {
    const keys = Array.isArray(path) ? path : path.split('.');
    const lastKey = keys.pop();
    let current = object;
    for (const key of keys) {
      if (!(key in current)) current[key] = {};
      current = current[key];
    }
    current[lastKey] = value;
    return object;
  },

  merge: function(...objects) {
    return Object.assign({}, ...objects);
  },

  pick: function(object, ...keys) {
    const result = {};
    for (const key of keys.flat()) {
      if (key in object) result[key] = object[key];
    }
    return result;
  },

  omit: function(object, ...keys) {
    const keysSet = new Set(keys.flat());
    const result = {};
    for (const key in object) {
      if (!keysSet.has(key)) result[key] = object[key];
    }
    return result;
  },

  // String methods
  camelCase: function(string) {
    return string.replace(/[-_\s]+(.)?/g, (_, c) => c ? c.toUpperCase() : '');
  },

  capitalize: function(string) {
    return string.charAt(0).toUpperCase() + string.slice(1).toLowerCase();
  },

  kebabCase: function(string) {
    return string.replace(/([a-z])([A-Z])/g, '$1-$2').toLowerCase();
  },

  snakeCase: function(string) {
    return string.replace(/([a-z])([A-Z])/g, '$1_$2').toLowerCase();
  },

  trim: function(string, chars) {
    return chars ? string.replace(new RegExp(`^[${chars}]+|[${chars}]+$`, 'g'), '') : string.trim();
  },

  truncate: function(string, options = {}) {
    const length = options.length || 30;
    const omission = options.omission || '...';
    return string.length > length ? string.substring(0, length - omission.length) + omission : string;
  },

  // Function methods
  debounce: function(func, wait, options = {}) {
    let timeout;
    return function(...args) {
      clearTimeout(timeout);
      timeout = setTimeout(() => func.apply(this, args), wait);
    };
  },

  throttle: function(func, wait, options = {}) {
    let timeout, previous = 0;
    return function(...args) {
      const now = Date.now();
      if (!previous && options.leading === false) previous = now;
      const remaining = wait - (now - previous);
      if (remaining <= 0 || remaining > wait) {
        clearTimeout(timeout);
        previous = now;
        func.apply(this, args);
      } else if (!timeout && options.trailing !== false) {
        timeout = setTimeout(() => {
          previous = options.leading === false ? 0 : Date.now();
          timeout = null;
          func.apply(this, args);
        }, remaining);
      }
    };
  },

  memoize: function(func, resolver) {
    const cache = new Map();
    return function(...args) {
      const key = resolver ? resolver.apply(this, args) : args[0];
      if (cache.has(key)) return cache.get(key);
      const result = func.apply(this, args);
      cache.set(key, result);
      return result;
    };
  },

  once: function(func) {
    let called = false;
    let result;
    return function(...args) {
      if (!called) {
        called = true;
        result = func.apply(this, args);
      }
      return result;
    };
  },

  // Utility methods
  identity: function(value) {
    return value;
  },

  noop: function() {},

  constant: function(value) {
    return function() { return value; };
  },

  times: function(n, iteratee) {
    const result = [];
    for (let i = 0; i < n; i++) {
      result.push(iteratee(i));
    }
    return result;
  },

  random: function(min, max) {
    if (max === undefined) { max = min; min = 0; }
    return Math.floor(Math.random() * (max - min + 1)) + min;
  },

  range: function(start, end, step = 1) {
    if (end === undefined) { end = start; start = 0; }
    const result = [];
    for (let i = start; i < end; i += step) {
      result.push(i);
    }
    return result;
  }
};

module.exports = lodash;
module.exports.default = lodash;
module.exports.debounce = lodash.debounce;
module.exports.throttle = lodash.throttle;
module.exports.chunk = lodash.chunk;
module.exports.uniq = lodash.uniq;
module.exports.get = lodash.get;
module.exports.set = lodash.set;
module.exports.merge = lodash.merge;
module.exports.cloneDeep = lodash.cloneDeep;
