// react/index.js
// Fake React implementation for testing
// This file simulates a moderately sized library

const ReactVersion = '18.0.0';

function createElement(type, props, ...children) {
  return {
    type,
    props: props || {},
    children: children.length ? children : null
  };
}

function Component(props) {
  this.props = props;
  this.state = {};
  this.setState = function(newState) {
    this.state = Object.assign({}, this.state, newState);
  };
}

function PureComponent(props) {
  Component.call(this, props);
}
PureComponent.prototype = Object.create(Component.prototype);

const useState = (initial) => {
  let state = initial;
  const setState = (newState) => { state = newState; };
  return [state, setState];
};

const useEffect = (fn, deps) => {
  fn();
};

const useContext = (context) => {
  return context._currentValue;
};

const useMemo = (fn, deps) => {
  return fn();
};

const useCallback = (fn, deps) => {
  return fn;
};

const useRef = (initial) => {
  return { current: initial };
};

const useReducer = (reducer, initialState) => {
  let state = initialState;
  const dispatch = (action) => { state = reducer(state, action); };
  return [state, dispatch];
};

const createContext = (defaultValue) => {
  return {
    Provider: function({ value, children }) {},
    Consumer: function({ children }) {},
    _currentValue: defaultValue
  };
};

const forwardRef = (render) => {
  return render;
};

const memo = (component) => {
  return component;
};

const lazy = (factory) => {
  return factory;
};

const Suspense = function({ fallback, children }) {};

const Fragment = function({ children }) {};

const StrictMode = function({ children }) {};

// Add some bulk to make the package more realistic
const internalHelpers = {
  reconciler: function() { /* complex reconciliation logic */ },
  fiber: function() { /* fiber architecture simulation */ },
  scheduler: function() { /* scheduling logic */ },
  hooks: function() { /* hooks implementation */ },
  events: function() { /* synthetic events */ }
};

// Export API
module.exports = {
  version: ReactVersion,
  createElement,
  Component,
  PureComponent,
  useState,
  useEffect,
  useContext,
  useMemo,
  useCallback,
  useRef,
  useReducer,
  createContext,
  forwardRef,
  memo,
  lazy,
  Suspense,
  Fragment,
  StrictMode,
  __SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED: internalHelpers
};
