// rxjs/index.js
// Fake RxJS implementation for testing

class Observable {
  constructor(subscribe) {
    this._subscribe = subscribe;
  }

  subscribe(observer) {
    if (typeof observer === 'function') {
      observer = { next: observer };
    }
    return this._subscribe(observer);
  }

  pipe(...operators) {
    return operators.reduce((prev, fn) => fn(prev), this);
  }

  static create(subscribe) {
    return new Observable(subscribe);
  }

  static of(...values) {
    return new Observable(observer => {
      values.forEach(v => observer.next(v));
      observer.complete && observer.complete();
    });
  }

  static from(iterable) {
    return new Observable(observer => {
      for (const value of iterable) {
        observer.next(value);
      }
      observer.complete && observer.complete();
    });
  }

  static interval(period) {
    return new Observable(observer => {
      let i = 0;
      const id = setInterval(() => observer.next(i++), period);
      return () => clearInterval(id);
    });
  }

  static timer(delay, period) {
    return new Observable(observer => {
      const id = setTimeout(() => {
        observer.next(0);
        if (period) {
          let i = 1;
          const intervalId = setInterval(() => observer.next(i++), period);
          return () => clearInterval(intervalId);
        }
      }, delay);
      return () => clearTimeout(id);
    });
  }

  static fromEvent(target, eventName) {
    return new Observable(observer => {
      const handler = event => observer.next(event);
      target.addEventListener(eventName, handler);
      return () => target.removeEventListener(eventName, handler);
    });
  }
}

class Subject extends Observable {
  constructor() {
    super(observer => {
      this.observers.push(observer);
      return () => {
        const index = this.observers.indexOf(observer);
        if (index >= 0) this.observers.splice(index, 1);
      };
    });
    this.observers = [];
  }

  next(value) {
    this.observers.forEach(o => o.next && o.next(value));
  }

  error(err) {
    this.observers.forEach(o => o.error && o.error(err));
  }

  complete() {
    this.observers.forEach(o => o.complete && o.complete());
  }
}

class BehaviorSubject extends Subject {
  constructor(initialValue) {
    super();
    this._value = initialValue;
  }

  next(value) {
    this._value = value;
    super.next(value);
  }

  getValue() {
    return this._value;
  }
}

class ReplaySubject extends Subject {
  constructor(bufferSize) {
    super();
    this.bufferSize = bufferSize || Infinity;
    this.buffer = [];
  }

  next(value) {
    this.buffer.push(value);
    if (this.buffer.length > this.bufferSize) {
      this.buffer.shift();
    }
    super.next(value);
  }

  subscribe(observer) {
    this.buffer.forEach(v => observer.next && observer.next(v));
    return super.subscribe(observer);
  }
}

// Operators
const map = fn => source => new Observable(observer =>
  source.subscribe({
    next: v => observer.next(fn(v)),
    error: e => observer.error && observer.error(e),
    complete: () => observer.complete && observer.complete()
  })
);

const filter = predicate => source => new Observable(observer =>
  source.subscribe({
    next: v => predicate(v) && observer.next(v),
    error: e => observer.error && observer.error(e),
    complete: () => observer.complete && observer.complete()
  })
);

const tap = fn => source => new Observable(observer =>
  source.subscribe({
    next: v => { fn(v); observer.next(v); },
    error: e => observer.error && observer.error(e),
    complete: () => observer.complete && observer.complete()
  })
);

const take = count => source => new Observable(observer => {
  let taken = 0;
  const sub = source.subscribe({
    next: v => {
      if (taken++ < count) {
        observer.next(v);
        if (taken === count) {
          observer.complete && observer.complete();
          sub && typeof sub === 'function' && sub();
        }
      }
    },
    error: e => observer.error && observer.error(e),
    complete: () => observer.complete && observer.complete()
  });
  return sub;
});

const debounceTime = delay => source => new Observable(observer => {
  let timeout;
  return source.subscribe({
    next: v => {
      clearTimeout(timeout);
      timeout = setTimeout(() => observer.next(v), delay);
    },
    error: e => observer.error && observer.error(e),
    complete: () => observer.complete && observer.complete()
  });
});

const mergeMap = fn => source => new Observable(observer =>
  source.subscribe({
    next: v => fn(v).subscribe(observer),
    error: e => observer.error && observer.error(e),
    complete: () => observer.complete && observer.complete()
  })
);

const switchMap = fn => source => {
  let innerSub;
  return new Observable(observer =>
    source.subscribe({
      next: v => {
        innerSub && typeof innerSub === 'function' && innerSub();
        innerSub = fn(v).subscribe(observer);
      },
      error: e => observer.error && observer.error(e),
      complete: () => observer.complete && observer.complete()
    })
  );
};

const catchError = fn => source => new Observable(observer =>
  source.subscribe({
    next: v => observer.next(v),
    error: e => fn(e, source).subscribe(observer),
    complete: () => observer.complete && observer.complete()
  })
);

module.exports = {
  Observable,
  Subject,
  BehaviorSubject,
  ReplaySubject,
  of: Observable.of.bind(Observable),
  from: Observable.from.bind(Observable),
  interval: Observable.interval.bind(Observable),
  timer: Observable.timer.bind(Observable),
  fromEvent: Observable.fromEvent.bind(Observable),
  // Operators
  map,
  filter,
  tap,
  take,
  debounceTime,
  mergeMap,
  switchMap,
  catchError
};
