// moment/moment.js
// Fake moment implementation for testing

function Moment(date) {
  this._date = date ? new Date(date) : new Date();
}

Moment.prototype.format = function(format) {
  format = format || 'YYYY-MM-DD';
  const d = this._date;
  return format
    .replace('YYYY', d.getFullYear())
    .replace('MM', String(d.getMonth() + 1).padStart(2, '0'))
    .replace('DD', String(d.getDate()).padStart(2, '0'))
    .replace('HH', String(d.getHours()).padStart(2, '0'))
    .replace('mm', String(d.getMinutes()).padStart(2, '0'))
    .replace('ss', String(d.getSeconds()).padStart(2, '0'));
};

Moment.prototype.add = function(amount, unit) {
  const d = new Date(this._date);
  switch(unit) {
    case 'days': d.setDate(d.getDate() + amount); break;
    case 'months': d.setMonth(d.getMonth() + amount); break;
    case 'years': d.setFullYear(d.getFullYear() + amount); break;
    case 'hours': d.setHours(d.getHours() + amount); break;
    case 'minutes': d.setMinutes(d.getMinutes() + amount); break;
    case 'seconds': d.setSeconds(d.getSeconds() + amount); break;
  }
  return new Moment(d);
};

Moment.prototype.subtract = function(amount, unit) {
  return this.add(-amount, unit);
};

Moment.prototype.isBefore = function(other) {
  return this._date < (other._date || new Date(other));
};

Moment.prototype.isAfter = function(other) {
  return this._date > (other._date || new Date(other));
};

Moment.prototype.diff = function(other, unit) {
  const diff = this._date - (other._date || new Date(other));
  switch(unit) {
    case 'seconds': return Math.floor(diff / 1000);
    case 'minutes': return Math.floor(diff / 60000);
    case 'hours': return Math.floor(diff / 3600000);
    case 'days': return Math.floor(diff / 86400000);
    default: return diff;
  }
};

Moment.prototype.startOf = function(unit) {
  const d = new Date(this._date);
  switch(unit) {
    case 'day':
      d.setHours(0, 0, 0, 0);
      break;
    case 'month':
      d.setDate(1);
      d.setHours(0, 0, 0, 0);
      break;
    case 'year':
      d.setMonth(0, 1);
      d.setHours(0, 0, 0, 0);
      break;
  }
  return new Moment(d);
};

Moment.prototype.endOf = function(unit) {
  const d = new Date(this._date);
  switch(unit) {
    case 'day':
      d.setHours(23, 59, 59, 999);
      break;
    case 'month':
      d.setMonth(d.getMonth() + 1, 0);
      d.setHours(23, 59, 59, 999);
      break;
    case 'year':
      d.setMonth(11, 31);
      d.setHours(23, 59, 59, 999);
      break;
  }
  return new Moment(d);
};

Moment.prototype.valueOf = function() {
  return this._date.getTime();
};

Moment.prototype.toDate = function() {
  return new Date(this._date);
};

function moment(date) {
  return new Moment(date);
}

moment.utc = function(date) {
  return new Moment(date);
};

moment.unix = function(timestamp) {
  return new Moment(timestamp * 1000);
};

moment.locale = function(locale) {
  return locale;
};

moment.duration = function(amount, unit) {
  return { _milliseconds: amount };
};

// Locale data (simulated)
const localeData = {
  months: ['January', 'February', 'March', 'April', 'May', 'June',
           'July', 'August', 'September', 'October', 'November', 'December'],
  weekdays: ['Sunday', 'Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday'],
  formats: {
    LT: 'h:mm A',
    LTS: 'h:mm:ss A',
    L: 'MM/DD/YYYY',
    LL: 'MMMM D, YYYY',
    LLL: 'MMMM D, YYYY h:mm A',
    LLLL: 'dddd, MMMM D, YYYY h:mm A'
  }
};

moment.localeData = function() {
  return localeData;
};

module.exports = moment;
module.exports.default = moment;
